databuilder/typing_utils.py

```diff
@@ -0,0 +1,143 @@
+"""
+This module contains some utilities for doing runtime type-checking. It is in no sense
+complete in that there are large areas of Python's type system which it doesn't attempt
+to handle. Its aim is to implement just enough behaviour to provide the validation
+needed by the Query Model.
+"""
+import typing
+from collections.abc import Mapping, Set
+from functools import singledispatch
+
+
+def get_typevars(typespec):
+    "Return set of all TypeVars nested anywhere inside `typespec`"
+    if isinstance(typespec, typing.TypeVar):
+        return {typespec}
+    else:
+        return set().union(*[get_typevars(arg) for arg in typing.get_args(typespec)])
+
+
+def type_matches(type_, spec, typevar_context):
+    """
+    Checks that the type given by `type_` matches the specification given by `spec`
+
+    For example:
+
+        `dict[str, FileNotFoundError]` matches `dict[Any, OSError]`
+
+    Because `str` matches `Any` and `FileNotFoundError` is a subclass of `OSError`.
+
+    The specification may contain TypeVars. The first match for each variable will bind
+    its value, recording it in the `typevar_context` dict. Subsequent uses of the
+    variable will fail to match if they don't have the same type. By sharing the
+    `typevar_context` dict across different calls to this function you can enforce
+    consistent interpretation of the TypeVar.
+
+    For example, both the below calls will match:
+
+        T = TypeVar("T")
+        type_matches(dict[int, int], dict[T, T], {})
+        type_matches(dict[str, str], dict[T, T], {})
+
+    But these will fail:
+
+        ctx = {}
+        type_matches(dict[int, bool], dict[T, T], ctx)
+        type_matches(dict[str, str], dict[T, T], ctx)
+
+    The first because the value of T is internally inconsistent; the second because
+    although it is internally consistent it is inconsistent with the previous example
+    whose context it shares.
+    """
+    # If `spec` is a type variable
+    if isinstance(spec, typing.TypeVar):
+        # Check the type constraints are met, if any
+        if spec.__constraints__:
+            if not any(
+                type_matches(type_, subspec, typevar_context)
+                for subspec in spec.__constraints__
+            ):
+                return False
+        # If we've already assigned a value for this variable (and it wasn't the Any
+        # type) then check it matches
+        if spec in typevar_context and typevar_context[spec] != typing.Any:
+            if typevar_context[spec] != type_:
+                return False
+        # Otherwise record this value as the value of the TypeVar
+        else:
+            typevar_context[spec] = type_
+        return True
+
+    # `Any` is the easy case: it just matches everything
+    if type_ is typing.Any or spec is typing.Any:
+        return True
+
+    spec_origin = typing.get_origin(spec)
+    spec_args = typing.get_args(spec)
```

Comment:
I was going to agree, but looking at in context now I'm not so sure. It would break up the `if .. elif .. else` block and I'd quite like to preserve the general structure of "get origin, then switch on origin type`. Maybe it would be clearer to just duplicate the `get_args` call inside the two blocks that need it?